DAS_Sistemas Develop Push: Práctica 1

daguirrez opened the pull request:
Práctica 1, incluido con un pequeño script que prueba la clase de matrices.

daguirrez committed:
Práctica 1 - Matrix

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,52 @@
+		except Exception:
+			pass
+
+	def getRow(self, index):
```
Te recomiendo echarle un ojo a los estándares oficiales de Python en cuanto al coding style: https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,52 @@
+			self.__matrix = list(
+				map(
+					lambda x: list(map(lambda y: float(y), x)),
+					map(lambda i: i.split(" "), matrix_str.split("\n"))
```
`split` por default toma funciones en blanco: https://docs.python.org/3.3/library/stdtypes.html?highlight=split#str.split

AnhellO replied:
Échale un ojo a https://docs.python.org/3/library/stdtypes.html#str.splitlines

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,52 @@
+
+	def getRow(self, index):
+		if index < len(self.__matrix):
+			return self.__matrix[index]
```
A pesar de que las listas son `0-index-based`, en los [specs de la práctica](https://github.com/AnhellO/DAS_Sistemas/blob/development/Ago-Dic-2020/Practicas/1er-Parcial/Pr%C3%A1ctica%201/SPECS.md) se hace uso de `1-index-based`

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,52 @@
+	def __init__(self, matrix_str):
+		try:
+			self.__matrix = list(
+				map(
```
Tu approach para solucionar esto es atractivo, sin embargo, es compleja, y creo que podría mantenerse más simple utilizando listas comprimidas.

Para describir mejor esto, veamos el siguiente escenario: alguien que quiera darle mantenimiento al código más adelante, ¿qué approach puede ser el más sencillo?, ¿cuál approach sería el más complicado?, ¿cuál puede ser el punto medio en donde encontremos un balance entre una solución atractiva, óptima, y legible para los demás developers que tuvieran que trabajar con tu código en un futuro?.

En caso de que el punto más complejo sea necesario ya que no hay otras alternativas, entonces optemos por también dejar documentado y descrito paso por paso cómo es que esto funciona, y de qué se encarga cada parte.

Te recomiendo darle un vistazo al siguiente artículo: https://softwarebrothers.co/blog/how-to-avoid-over-engineering/, y por supuesto, importante darle también un vistazo al [zen de python](https://www.python.org/dev/peps/pep-0020/#id2).

Por supuesto, esto funciona, y es opcional cambiar tu approach, solamente hago este comentario como sugerencia y para que puedas tomarlo en cuenta para un futuro 😄

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,52 @@
+		return []
+
+	def getColumn(self, index):
+		try:
```
Creo que me faltó poner en los specs que podían asumir que es siempre seguro recibir un índice dentro de los límites de la matriz, así como el hecho de que se puede asumir que siempre recibirás un `string` que represente una matriz válida, una disculpa 😅 . Puedes remover o dejar los bloques `try-except` dentro del código, ya lo dejo a tu criterio 👍

AnhellO reviewed (changes requested):
Dejé algunos comentarios @daguirrez, avísame si tienes alguna duda 👍

daguirrez committed:
Práctica 1 - Correcciones y optimizaciones.

- Ahora los nombres de los métodos siguen la convención de nombres en
python.
- Los indices de las filas y columnas son 1-index-based.
- Optimizaciones en base a entradas de indice seguras.
- Optimización en la creación de la matriz en base a una cadena segura.

AnhellO commented on Ago-Dic-2020/aguirre-juarez-david/practica-1/matrix.py:
```diff
@@ -0,0 +1,35 @@
+class Matrix:
+	__matrix = []
```
No es necesario declarar la matriz acá. Esto debido a que queremos que sea variable de instancia y no de clase: https://www.digitalocean.com/community/tutorials/understanding-class-and-instance-variables-in-python-3

AnhellO reviewed (changes requested):
Muy bien @daguirrez!, Un pequeño comentario más y estamos listos 😉

daguirrez committed:
Práctica 1 - Corrección de tipo de atributo __matrix
